tests/TestHarness/Cluster.py

```diff
@@ -993,34 +995,33 @@ def parseClusterKeys(totalNodes):
         Utils.Print(f'Found {len(producerKeys)} producer keys')
         return producerKeys
 
-    def activateInstantFinality(self, launcher, biosFinalizer, pnodes):
+    def activateInstantFinality(self, biosFinalizer=True):
         # call setfinalizer
         numFins = 0
-        for n in launcher.network.nodes.values():
-            if not n.keys or not n.keys[0].blspubkey:
+        for n in (self.nodes + [self.biosNode]):
+            if not n or not n.keys or not n.keys[0].blspubkey:
                 continue
-            if not n.producers:
+            if not n.isProducer:
                 continue
-            if n.index == Node.biosNodeId and not biosFinalizer:
+            if n.nodeId == 'bios' and not biosFinalizer:
                 continue
             numFins = numFins + 1
 
         threshold = int(numFins * 2 / 3 + 1)
         if threshold > 2 and threshold == numFins:
             # nodes are often stopped, so do not require all node votes
             threshold = threshold - 1
-        # pnodes does not include biosNode
-        if Utils.Debug: Utils.Print(f"threshold: {threshold}, numFins: {numFins}, pnodes: {pnodes}")
+        if Utils.Debug: Utils.Print(f"threshold: {threshold}, numFins: {numFins}")
         setFinStr =  f'{{"finalizer_policy": {{'
         setFinStr += f'  "threshold": {threshold}, '
         setFinStr += f'  "finalizers": ['
         finNum = 1
-        for n in launcher.network.nodes.values():
-            if n.index == Node.biosNodeId and not biosFinalizer:
+        for n in (self.nodes + [self.biosNode]):
```

Comment:
Can this for loop be combined with the one above? They are in similar shape.

Reply:
I don't think so, need to have `threshold` calculated first.